tests/models/decoder_only/language/test_bamba.py

```diff
@@ -0,0 +1,329 @@
+import pytest
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+from vllm.config import VllmConfig
+from vllm.sampling_params import SamplingParams
+
+from ...utils import check_outputs_equal
+
+# will be ch
+MODELS = ["ibm-fms/Bamba-9.8b-1.8T-hf"]
+
+
+# Use lower-level interfaces to create this greedy generator, as mamba will
+# choke on the model_kwarg 'attention_mask' if hf_model.generate_greedy is used.
```

Comment:
Is this actually true of Bamba? I think this might not be the case for hybrid models, only attention-free ones.

We could consider renaming `tests/models/decoder_only/language/test_jamba.py` to `tests/models/decoder_only/language/test_hybrid.py` and then extending it to support bamba as well. Alternatively we could adapt the `test_jamba.py` tests to test bamba instead of adapting `test_mamba.py` tests.

Reply:
@tlrmchlsmth ah ok.. thats a good point.. let me look at `test_jamba.py`

Reply:
I did a quick test on the `test_jamba.py`, everything passed except these 3 below
- as expected prefill chunking fails for now until the WIP is resolved.
- `test_parallel_sampling` we seem to require `float` to pass the similar test in `test_bamba.py`
- for the last one Im abit confused. the stacktrace was `Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method`

<details>
Traceback (most recent call last):
  File "/data/flim/vllm/tests/utils.py", line 658, in wrapper
    f(*args, **kwargs)
  File "/data/flim/vllm/tests/models/decoder_only/language/test_jamba.py", line 328, in test_jamba_distributed_produces_identical_generation
    with vllm_runner(model, dtype=dtype, tensor_parallel_size=2) as vllm_model:
  File "/data/flim/vllm/tests/conftest.py", line 672, in __init__
    self.model = LLM(
  File "/data/flim/vllm/vllm/utils.py", line 1063, in inner
    return fn(*args, **kwargs)
  File "/data/flim/vllm/vllm/entrypoints/llm.py", line 232, in __init__
    self.llm_engine = self.engine_class.from_engine_args(
  File "/data/flim/vllm/vllm/engine/llm_engine.py", line 575, in from_engine_args
    engine = cls(
  File "/data/flim/vllm/vllm/engine/llm_engine.py", line 336, in __init__
    self.model_executor = executor_class(vllm_config=vllm_config, )
  File "/data/flim/vllm/vllm/executor/distributed_gpu_executor.py", line 26, in __init__
    super().__init__(*args, **kwargs)
  File "/data/flim/vllm/vllm/executor/executor_base.py", line 36, in __init__
    self._init_executor()
  File "/data/flim/vllm/vllm/executor/multiproc_gpu_executor.py", line 113, in _init_executor
    self._run_workers("init_device")
  File "/data/flim/vllm/vllm/executor/multiproc_gpu_executor.py", line 195, in _run_workers
    driver_worker_output = driver_worker_method(*args, **kwargs)
  File "/data/flim/vllm/vllm/worker/worker.py", line 138, in init_device
    torch.cuda.set_device(self.device)
  File "/workspace/mamba-vllm/lib/python3.10/site-packages/torch/cuda/__init__.py", line 478, in set_device
    torch._C._cuda_setDevice(device)
  File "/workspace/mamba-vllm/lib/python3.10/site-packages/torch/cuda/__init__.py", line 305, in _lazy_init
    raise RuntimeError(
RuntimeError: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method

</details>

```
FAILED tests/models/decoder_only/language/test_jamba.py::test_mamba_prefill_chunking[10-bfloat16-ibm-fms/Bamba-9.8b-1.8T-hf] - AssertionError: Test0:
FAILED tests/models/decoder_only/language/test_jamba.py::test_parallel_sampling[15-bfloat16-ibm-fms/Bamba-9.8b-1.8T-hf] - AssertionError: Test0:
FAILED tests/models/decoder_only/language/test_jamba.py::test_jamba_distributed_produces_identical_generation[64-float-ibm-fms/Bamba-9.8b-1.8T-hf] - AssertionError: function <function test_jamba_distributed_produces_identical_generation at 0x7fbefda9a5f0> failed when called with args () and kwargs {'vllm_runner': <class 'tests.conftest.VllmRunner'>, 'model': 'i...
```

Reply:
Ah ok for multi-gpu tests you need to do `pytest -v -s` to avoid the fork error. The test passes!

<details>

Fork a new process to run a test 0
INFO 12-13 02:08:38 __init__.py:46] No plugins found.
INFO 12-13 02:08:55 config.py:403] This model supports multiple tasks: {'generate', 'embedding'}. Defaulting to 'generate'.
INFO 12-13 02:08:55 config.py:1038] Defaulting to use mp for distributed inference
INFO 12-13 02:08:55 llm_engine.py:249] Initializing an LLM engine (v0.1.dev3512+g388ee3d) with config: model='ibm-fms/Bamba-9.8b-1.8T-hf', speculative_config=None, tokenizer='ibm-fms/Bamba-9.8b-1.8T-hf', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.float32, max_seq_len=1024, download_dir=None, load_format=LoadFormat.AUTO, tensor_parallel_size=2, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=False, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='xgrammar'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None, collect_model_forward_time=False, collect_model_execute_time=False), seed=0, served_model_name=ibm-fms/Bamba-9.8b-1.8T-hf, num_scheduler_steps=1, chunked_prefill_enabled=False multi_step_stream_outputs=True, enable_prefix_caching=False, use_async_output_proc=True, use_cached_outputs=False, mm_processor_kwargs=None, pooler_config=None,compilation_config=CompilationConfig(level=0, backend='', custom_ops=[], splitting_ops=['vllm.unified_attention', 'vllm.unified_attention_with_output'], use_inductor=True, inductor_specialize_for_cudagraph_no_more_than=None, inductor_compile_sizes=[], inductor_compile_config={}, inductor_passes={}, use_cudagraph=False, cudagraph_num_of_warmups=0, cudagraph_capture_sizes=None, cudagraph_copy_inputs=False, pass_config=PassConfig(dump_graph_stages=[], dump_graph_dir=PosixPath('.'), enable_fusion=True, enable_reshape=True), compile_sizes=[], capture_sizes=[256, 248, 240, 232, 224, 216, 208, 200, 192, 184, 176, 168, 160, 152, 144, 136, 128, 120, 112, 104, 96, 88, 80, 72, 64, 56, 48, 40, 32, 24, 16, 8, 4, 2, 1], enabled_custom_ops=Counter(), disabled_custom_ops=Counter(), static_forward_context={})
WARNING 12-13 02:09:01 multiproc_gpu_executor.py:56] Reducing Torch parallelism from 40 threads to 1 to avoid unnecessary CPU contention. Set OMP_NUM_THREADS in the external environment to tune this value as needed.
INFO 12-13 02:09:01 custom_cache_manager.py:17] Setting Triton cache manager to: vllm.triton_utils.custom_cache_manager:CustomCacheManager
INFO 12-13 02:09:01 selector.py:222] Cannot use FlashAttention-2 backend for dtype other than torch.float16 or torch.bfloat16.
INFO 12-13 02:09:01 selector.py:129] Using XFormers backend.
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:01 selector.py:222] Cannot use FlashAttention-2 backend for dtype other than torch.float16 or torch.bfloat16.
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:01 selector.py:129] Using XFormers backend.
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:01 multiproc_worker_utils.py:215] Worker ready; awaiting tasks
INFO 12-13 02:09:02 utils.py:996] Found nccl from library libnccl.so.2
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:02 utils.py:996] Found nccl from library libnccl.so.2
INFO 12-13 02:09:02 pynccl.py:69] vLLM is using nccl==2.21.5
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:02 pynccl.py:69] vLLM is using nccl==2.21.5
INFO 12-13 02:09:02 custom_all_reduce_utils.py:242] reading GPU P2P access cache from /workspace/.cache/vllm/gpu_p2p_access_cache_for_0,1.json
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:02 custom_all_reduce_utils.py:242] reading GPU P2P access cache from /workspace/.cache/vllm/gpu_p2p_access_cache_for_0,1.json
INFO 12-13 02:09:02 shm_broadcast.py:236] vLLM message queue communication handle: Handle(connect_ip='127.0.0.1', local_reader_ranks=[1], buffer=<vllm.distributed.device_communicators.shm_broadcast.ShmRingBuffer object at 0x7f940c04f730>, local_subscribe_port=33527, remote_subscribe_port=None)
INFO 12-13 02:09:02 model_runner.py:1089] Starting to load model ibm-fms/Bamba-9.8b-1.8T-hf...
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:02 model_runner.py:1089] Starting to load model ibm-fms/Bamba-9.8b-1.8T-hf...
INFO 12-13 02:09:06 weight_utils.py:243] Using model weights format ['*.safetensors']
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:06 weight_utils.py:243] Using model weights format ['*.safetensors']
Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]
Loading safetensors checkpoint shards:  25% Completed | 1/4 [00:01<00:04,  1.40s/it]
Loading safetensors checkpoint shards:  50% Completed | 2/4 [00:02<00:02,  1.49s/it]
Loading safetensors checkpoint shards:  75% Completed | 3/4 [00:04<00:01,  1.53s/it]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:06<00:00,  1.57s/it]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:06<00:00,  1.54s/it]

INFO 12-13 02:09:14 model_runner.py:1094] Loading model weights took 18.4902 GB
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:16 model_runner.py:1094] Loading model weights took 18.4902 GB
(VllmWorkerProcess pid=1587462) INFO 12-13 02:10:28 worker.py:237] Memory profiling results: duration=71.88 seconds, total_gpu_memory=79.15GiB, initial_memory_usage=19.68GiB, peak_torch_memory=34.32GiB, memory_usage_post_profile=35.32GiB, non_torch_memory=1.96GiB, kv_cache_size=34.96GiB, gpu_memory_utilization=0.90.
INFO 12-13 02:10:28 worker.py:237] Memory profiling results: duration=72.36 seconds, total_gpu_memory=79.15GiB, initial_memory_usage=19.68GiB, peak_torch_memory=34.50GiB, memory_usage_post_profile=35.32GiB, non_torch_memory=1.96GiB, kv_cache_size=34.78GiB, gpu_memory_utilization=0.90.
INFO 12-13 02:10:29 distributed_gpu_executor.py:58] # GPU blocks: 189945, # CPU blocks: 21845
INFO 12-13 02:10:29 distributed_gpu_executor.py:62] Maximum concurrency for 1024 tokens per request: 2967.89x
INFO 12-13 02:10:32 model_runner.py:1409] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.
INFO 12-13 02:10:32 model_runner.py:1413] If out-of-memory error occurs during cudagraph capture, consider decreasing `gpu_memory_utilization` or switching to eager mode. You can also reduce the `max_num_seqs` as needed to decrease memory usage.
(VllmWorkerProcess pid=1587462) INFO 12-13 02:10:32 model_runner.py:1409] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.
(VllmWorkerProcess pid=1587462) INFO 12-13 02:10:32 model_runner.py:1413] If out-of-memory error occurs during cudagraph capture, consider decreasing `gpu_memory_utilization` or switching to eager mode. You can also reduce the `max_num_seqs` as needed to decrease memory usage.
INFO 12-13 02:10:57 custom_all_reduce.py:224] Registering 3232 cuda graph addresses
(VllmWorkerProcess pid=1587462) INFO 12-13 02:10:57 custom_all_reduce.py:224] Registering 3232 cuda graph addresses
(VllmWorkerProcess pid=1587462) INFO 12-13 02:10:57 model_runner.py:1523] Graph capturing finished in 25 secs, took 1.56 GiB
INFO 12-13 02:10:57 model_runner.py:1523] Graph capturing finished in 25 secs, took 1.56 GiB
Processed prompts: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:06<00:00,  1.29it/s, est. speed input: 25.33 toks/s, output: 82.59 toks/s]
INFO 12-13 02:11:06 multiproc_worker_utils.py:133] Terminating local vLLM worker processes
(VllmWorkerProcess pid=1587462) INFO 12-13 02:11:06 multiproc_worker_utils.py:240] Worker exiting
INFO 12-13 02:11:20 config.py:403] This model supports multiple tasks: {'generate', 'embedding'}. Defaulting to 'generate'.
INFO 12-13 02:11:20 llm_engine.py:249] Initializing an LLM engine (v0.1.dev3512+g388ee3d) with config: model='ibm-fms/Bamba-9.8b-1.8T-hf', speculative_config=None, tokenizer='ibm-fms/Bamba-9.8b-1.8T-hf', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.float32, max_seq_len=1024, download_dir=None, load_format=LoadFormat.AUTO, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=False, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='xgrammar'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None, collect_model_forward_time=False, collect_model_execute_time=False), seed=0, served_model_name=ibm-fms/Bamba-9.8b-1.8T-hf, num_scheduler_steps=1, chunked_prefill_enabled=False multi_step_stream_outputs=True, enable_prefix_caching=False, use_async_output_proc=True, use_cached_outputs=False, mm_processor_kwargs=None, pooler_config=None,compilation_config=CompilationConfig(level=0, backend='', custom_ops=[], splitting_ops=['vllm.unified_attention', 'vllm.unified_attention_with_output'], use_inductor=True, inductor_specialize_for_cudagraph_no_more_than=None, inductor_compile_sizes=[], inductor_compile_config={}, inductor_passes={}, use_cudagraph=False, cudagraph_num_of_warmups=0, cudagraph_capture_sizes=None, cudagraph_copy_inputs=False, pass_config=PassConfig(dump_graph_stages=[], dump_graph_dir=PosixPath('.'), enable_fusion=True, enable_reshape=True), compile_sizes=[], capture_sizes=[256, 248, 240, 232, 224, 216, 208, 200, 192, 184, 176, 168, 160, 152, 144, 136, 128, 120, 112, 104, 96, 88, 80, 72, 64, 56, 48, 40, 32, 24, 16, 8, 4, 2, 1], enabled_custom_ops=Counter(), disabled_custom_ops=Counter(), static_forward_context={})
INFO 12-13 02:11:26 model_runner.py:1089] Starting to load model ibm-fms/Bamba-9.8b-1.8T-hf...
INFO 12-13 02:11:28 weight_utils.py:243] Using model weights format ['*.safetensors']
Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]
Loading safetensors checkpoint shards:  25% Completed | 1/4 [00:02<00:08,  2.85s/it]
Loading safetensors checkpoint shards:  50% Completed | 2/4 [00:05<00:05,  2.91s/it]
Loading safetensors checkpoint shards:  75% Completed | 3/4 [00:08<00:02,  2.91s/it]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:11<00:00,  2.92s/it]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:11<00:00,  2.91s/it]

INFO 12-13 02:11:42 model_runner.py:1094] Loading model weights took 36.6217 GB
INFO 12-13 02:11:45 worker.py:237] Memory profiling results: duration=2.65 seconds, total_gpu_memory=79.15GiB, initial_memory_usage=39.37GiB, peak_torch_memory=67.84GiB, memory_usage_post_profile=69.07GiB, non_torch_memory=2.73GiB, kv_cache_size=0.66GiB, gpu_memory_utilization=0.90.
INFO 12-13 02:11:45 gpu_executor.py:76] # GPU blocks: 1797, # CPU blocks: 10922
INFO 12-13 02:11:45 gpu_executor.py:80] Maximum concurrency for 1024 tokens per request: 28.08x
INFO 12-13 02:11:45 model_runner.py:1409] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.
INFO 12-13 02:11:45 model_runner.py:1413] If out-of-memory error occurs during cudagraph capture, consider decreasing `gpu_memory_utilization` or switching to eager mode. You can also reduce the `max_num_seqs` as needed to decrease memory usage.
INFO 12-13 02:12:08 model_runner.py:1523] Graph capturing finished in 23 secs, took 0.11 GiB
Processed prompts: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:01<00:00,  4.21it/s, est. speed input: 82.65 toks/s, output: 269.54 toks/s]
PASSED

==================================================================================================== warnings summary =====================================================================================================
../transformers/src/transformers/utils/hub.py:128
  /data/flim/transformers/src/transformers/utils/hub.py:128: FutureWarning: Using `TRANSFORMERS_CACHE` is deprecated and will be removed in v5 of Transformers. Use `HF_HOME` instead.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================================================== 1 passed, 1 warning in 216.66s (0:03:36) =========================================================================================
(mamba-vllm) 1000960000@flim-mamba-master-0:~/data/vllm$

</details>